pkg/adapter/gcppubsub/adapter.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcppubsub
+
+import (
+	"fmt"
+	"github.com/knative/pkg/logging"
+	"go.uber.org/zap"
+	"io/ioutil"
+	"net/http"
+
+	// Imports the Google Cloud Pub/Sub client package.
+	"cloud.google.com/go/pubsub"
+	"github.com/knative/pkg/cloudevents"
+	"golang.org/x/net/context"
+)
+
+const (
+	eventType = "google.pubsub.topic.publish"
+)
+
+// Adapter implements the GCP Pub/Sub adapter to deliver Pub/Sub messages from
+// a pre-existing topic/subscription to a Sink.
+type Adapter struct {
+	// ProjectID is the pre-existing gcp project id to use.
+	ProjectID string
+	// TopicID is the pre-existing gcp pub/sub topic id to use.
+	TopicID string
+	// SubscriptionID is the pre-existing gcp pub/sub subscription id to use.
+	SubscriptionID string
+	// SinkURI is the URI messages will be forwarded on to.
+	SinkURI string
+
+	source       string
+	client       *pubsub.Client
+	subscription *pubsub.Subscription
+}
+
+func (a *Adapter) Start(ctx context.Context) error {
+	a.source = fmt.Sprintf("//pubsub.googleapis.com/%s/topics/%s", a.ProjectID, a.TopicID)
+
+	var err error
+	// Make the client to pubsub
+	if a.client, err = pubsub.NewClient(ctx, a.ProjectID); err != nil {
+		return err
+	}
+
+	// Set the subscription from the client
+	a.subscription = a.client.Subscription(a.SubscriptionID)
+
+	// Using that subscription, start receiving messages.
+	if err := a.subscription.Receive(ctx, func(ctx context.Context, m *pubsub.Message) {
+		a.receiveMessage(ctx, &PubSubMessageWrapper{M: m})
+		m.Nack()
```

Comment:
nit: I found this little hard to read with ack/nack handling in the receivemessage too with the anonymous function here. but it's fine.